pkg/azuredisk/controllerserver.go

```diff
@@ -1239,6 +1254,183 @@ func (d *Driver) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsReques
 	return azureutils.GetEntriesAndNextToken(req, snapshots)
 }
 
+func (d *Driver) CreateVolumeGroupSnapshot(ctx context.Context, req *csi.CreateVolumeGroupSnapshotRequest) (*csi.CreateVolumeGroupSnapshotResponse, error) {
+	sourceVolumeIDs := req.GetSourceVolumeIds()
+	if len(sourceVolumeIDs) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "CreateVolumeGroupSnapshot Source Volume IDs must be provided")
+	}
+	for _, sourceVolumeID := range sourceVolumeIDs {
+		if len(sourceVolumeID) == 0 {
+			return nil, status.Error(codes.InvalidArgument, "CreateVolumeGroupSnapshot Source Volume ID must be provided")
+		}
+	}
+	volumeGroupSnapshotName := req.Name
+	if len(volumeGroupSnapshotName) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "volume group snapshot name must be provided")
+	}
+
+	snapshotNames := []string{}
+	for _, sourceVolumeID := range sourceVolumeIDs {
+		groupSnapshotUUID := strings.TrimPrefix(volumeGroupSnapshotName, "groupsnapshot-")
+		snapshotName := fmt.Sprintf("snapshot-%x", sha256.Sum256([]byte(groupSnapshotUUID+sourceVolumeID)))
+		snapshotName = azureutils.CreateValidDiskName(snapshotName)
+		snapshotNames = append(snapshotNames, snapshotName)
+	}
+
+	if acquired := d.volumeLocks.TryAcquire(volumeGroupSnapshotName); !acquired {
+		return nil, status.Errorf(codes.Aborted, volumeOperationAlreadyExistsFmt, volumeGroupSnapshotName)
+	}
+	defer d.volumeLocks.Release(volumeGroupSnapshotName)
+
+	metricsRequest := "controller_create_volume_group_snapshot"
+
+	mc := metrics.NewMetricContext(consts.AzureDiskCSIDriverName, metricsRequest, d.cloud.ResourceGroup, d.cloud.SubscriptionID, d.Name)
+	isOperationSucceeded := false
+	defer func() {
+		mc.ObserveOperationWithResult(isOperationSucceeded, consts.GroupSnapshotID, volumeGroupSnapshotName)
+		for idx, sourceVolumeID := range sourceVolumeIDs {
+			mc.ObserveOperationWithResult(isOperationSucceeded, consts.SourceResourceID, sourceVolumeID, consts.SnapshotName, snapshotNames[idx])
```

Comment:
not necessary metrics here since we already have metrics in every CreateSnapshot call

Reply:
done